File: backend/routes/gst_invoice.py
"""GST Invoice routes - CRUD + PDF + WhatsApp"""
from fastapi import APIRouter, HTTPException
from database import db
from typing import Optional, List
from pydantic import BaseModel
from datetime import datetime, timezone
import uuid, io
from reportlab.lib.pagesizes import A4
from reportlab.lib import colors
from reportlab.lib.units import mm
from reportlab.platypus import SimpleDocTemplate, Table as RTable, TableStyle, Paragraph, Spacer, HRFlowable
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from starlette.responses import StreamingResponse
import os

router = APIRouter()

# Register Hindi fonts
font_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'fonts')
for fname, ffile in [('FreeSans', 'FreeSans.ttf'), ('FreeSansBold', 'FreeSansBold.ttf')]:
    fpath = os.path.join(font_dir, ffile)
    if os.path.exists(fpath):
        try:
            pdfmetrics.registerFont(TTFont(fname, fpath))
        except Exception:
            pass

# ============ GST COMPANY SETTINGS ============

class GstCompanySettings(BaseModel):
    company_name: str = ""
    gstin: str = ""
    address: str = ""
    state_code: str = ""
    state_name: str = ""
    phone: str = ""
    bank_name: str = ""
    bank_account: str = ""
    bank_ifsc: str = ""

@router.get("/gst-company-settings")
async def get_gst_company_settings():
    s = await db.settings.find_one({"key": "gst_company"}, {"_id": 0})
    if not s:
        return {"company_name": "", "gstin": "", "address": "", "state_code": "21", "state_name": "Odisha", "phone": "", "bank_name": "", "bank_account": "", "bank_ifsc": ""}
    return {k: s.get(k, "") for k in ["company_name", "gstin", "address", "state_code", "state_name", "phone", "bank_name", "bank_account", "bank_ifsc"]}

@router.put("/gst-company-settings")
async def update_gst_company_settings(data: GstCompanySettings):
    await db.settings.update_one(
        {"key": "gst_company"},
        {"$set": {"key": "gst_company", **data.dict()}},
        upsert=True
    )
    return {"success": True}

# ============ GST INVOICE CRUD ============

class GstInvoiceItem(BaseModel):
    name: str = ""
    hsn: str = ""
    qty: float = 0
    unit: str = "QNTL"
    rate: float = 0
    gst_pct: float = 5

class GstInvoiceCreate(BaseModel):
    invoice_no: str
    date: str
    buyer_name: str = ""
    buyer_gstin: str = ""
    buyer_address: str = ""
    buyer_phone: str = ""
    is_igst: bool = False
    items: List[GstInvoiceItem] = []
    kms_year: str = ""
    season: str = ""
    notes: str = ""

@router.get("/gst-invoices")
async def list_gst_invoices(kms_year: str = "", season: str = ""):
    filt = {}
    if kms_year:
        filt["kms_year"] = kms_year
    if season:
        filt["season"] = season
    docs = await db.gst_invoices.find(filt, {"_id": 0}).sort("date", -1).to_list(5000)
    return docs

@router.post("/gst-invoices")
async def create_gst_invoice(data: GstInvoiceCreate):
    inv = data.dict()
    inv["id"] = str(uuid.uuid4())
    inv["created_at"] = datetime.now(timezone.utc).isoformat()
    # Calculate totals
    taxable = sum((it["qty"] or 0) * (it["rate"] or 0) for it in inv["items"])
    gst_amt = sum((it["qty"] or 0) * (it["rate"] or 0) * (it["gst_pct"] or 0) / 100 for it in inv["items"])
    inv["totals"] = {
        "taxable": round(taxable, 2),
        "gst": round(gst_amt, 2),
        "cgst": round(gst_amt / 2, 2) if not inv["is_igst"] else 0,
        "sgst": round(gst_amt / 2, 2) if not inv["is_igst"] else 0,
        "igst": round(gst_amt, 2) if inv["is_igst"] else 0,
        "total": round(taxable + gst_amt, 2)
    }
    await db.gst_invoices.insert_one(inv)
    inv.pop("_id", None)
    return inv

@router.put("/gst-invoices/{inv_id}")
async def update_gst_invoice(inv_id: str, data: GstInvoiceCreate):
    inv = data.dict()
    taxable = sum((it["qty"] or 0) * (it["rate"] or 0) for it in inv["items"])
    gst_amt = sum((it["qty"] or 0) * (it["rate"] or 0) * (it["gst_pct"] or 0) / 100 for it in inv["items"])
    inv["totals"] = {
        "taxable": round(taxable, 2),
        "gst": round(gst_amt, 2),
        "cgst": round(gst_amt / 2, 2) if not inv["is_igst"] else 0,
        "sgst": round(gst_amt / 2, 2) if not inv["is_igst"] else 0,
        "igst": round(gst_amt, 2) if inv["is_igst"] else 0,
        "total": round(taxable + gst_amt, 2)
    }
    inv["updated_at"] = datetime.now(timezone.utc).isoformat()
    result = await db.gst_invoices.update_one({"id": inv_id}, {"$set": inv})
    if result.matched_count == 0:
        raise HTTPException(404, "Invoice not found")
    return {"success": True, "totals": inv["totals"]}

@router.delete("/gst-invoices/{inv_id}")
async def delete_gst_invoice(inv_id: str):
    result = await db.gst_invoices.delete_one({"id": inv_id})
    if result.deleted_count == 0:
        raise HTTPException(404, "Invoice not found")
    return {"success": True}

# ============ GST INVOICE PDF ============

def _fmt(v):
    return f"{v:,.2f}" if v else "0.00"

@router.get("/gst-invoices/{inv_id}/pdf")
async def gst_invoice_pdf(inv_id: str):
    inv = await db.gst_invoices.find_one({"id": inv_id}, {"_id": 0})
    if not inv:
        raise HTTPException(404, "Invoice not found")
    company = await db.settings.find_one({"key": "gst_company"}, {"_id": 0}) or {}

    buf = io.BytesIO()
    doc = SimpleDocTemplate(buf, pagesize=A4, leftMargin=20*mm, rightMargin=20*mm, topMargin=15*mm, bottomMargin=15*mm)
    styles = getSampleStyleSheet()

    # Font selection
    try:
        pdfmetrics.getFont('FreeSansBold')
        bfont = 'FreeSansBold'
        nfont = 'FreeSans'
    except Exception:
        bfont = 'Helvetica-Bold'
        nfont = 'Helvetica'

    title_style = ParagraphStyle('title', fontName=bfont, fontSize=16, alignment=1, spaceAfter=2)
    sub_style = ParagraphStyle('sub', fontName=nfont, fontSize=9, alignment=1, spaceAfter=1, textColor=colors.grey)
    label_style = ParagraphStyle('label', fontName=nfont, fontSize=8, textColor=colors.grey)
    val_style = ParagraphStyle('val', fontName=bfont, fontSize=9)
    small_style = ParagraphStyle('small', fontName=nfont, fontSize=8)

    elements = []

    # === HEADER ===
    elements.append(Paragraph(company.get("company_name", "COMPANY NAME"), title_style))
    if company.get("address"):
        elements.append(Paragraph(company["address"], sub_style))
    if company.get("gstin"):
        elements.append(Paragraph(f"GSTIN: {company['gstin']}", sub_style))
    if company.get("phone"):
        elements.append(Paragraph(f"Phone: {company['phone']}", sub_style))
    elements.append(Spacer(1, 4))
    elements.append(HRFlowable(width="100%", thickness=1, color=colors.HexColor('#334155')))
    elements.append(Spacer(1, 4))

    # TAX INVOICE title
    elements.append(Paragraph("<b>TAX INVOICE</b>", ParagraphStyle('inv', fontName=bfont, fontSize=14, alignment=1, textColor=colors.HexColor('#1e40af'))))
    elements.append(Spacer(1, 6))

    # Invoice info + Buyer info in a 2-column table
    info_data = [
        [Paragraph(f"<b>Invoice No:</b> {inv.get('invoice_no','')}", small_style),
         Paragraph(f"<b>Date:</b> {inv.get('date','')}", small_style)],
        [Paragraph(f"<b>Bill To:</b> {inv.get('buyer_name','')}", val_style),
         Paragraph(f"<b>GSTIN:</b> {inv.get('buyer_gstin','')}", small_style)],
    ]
    if inv.get('buyer_address'):
        info_data.append([Paragraph(f"Address: {inv['buyer_address']}", small_style), Paragraph("", small_style)])
    info_t = RTable(info_data, colWidths=[280, 280])
    info_t.setStyle(TableStyle([
        ('VALIGN', (0, 0), (-1, -1), 'TOP'),
        ('TOPPADDING', (0, 0), (-1, -1), 2),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 2),
    ]))
    elements.append(info_t)
    elements.append(Spacer(1, 8))

    # === ITEMS TABLE ===
    items = inv.get("items", [])
    hdr = ['#', 'Item', 'HSN', 'Qty', 'Unit', 'Rate', 'Taxable', 'GST%', 'GST Amt', 'Total']
    rows = [hdr]
    for i, it in enumerate(items):
        taxable = (it.get("qty", 0) or 0) * (it.get("rate", 0) or 0)
        gst_a = taxable * (it.get("gst_pct", 0) or 0) / 100
        total = taxable + gst_a
        rows.append([
            str(i + 1), it.get("name", ""), it.get("hsn", ""),
            str(round(it.get("qty", 0), 2)), it.get("unit", ""),
            f"Rs.{_fmt(it.get('rate', 0))}", f"Rs.{_fmt(taxable)}",
            f"{it.get('gst_pct', 0)}%", f"Rs.{_fmt(gst_a)}", f"Rs.{_fmt(total)}"
        ])

    totals = inv.get("totals", {})
    rows.append(['', '', '', '', '', '', f"Rs.{_fmt(totals.get('taxable', 0))}", '',
                 f"Rs.{_fmt(totals.get('gst', 0))}", f"Rs.{_fmt(totals.get('total', 0))}"])

    col_w = [22, 95, 55, 40, 35, 55, 60, 35, 50, 60]
    t = RTable(rows, colWidths=col_w)
    t_style = [
        ('FONTNAME', (0, 0), (-1, 0), bfont),
        ('FONTNAME', (0, 1), (-1, -2), nfont),
        ('FONTNAME', (0, -1), (-1, -1), bfont),
        ('FONTSIZE', (0, 0), (-1, -1), 7.5),
        ('BACKGROUND', (0, 0), (-1, 0), colors.HexColor('#e2e8f0')),
        ('BACKGROUND', (0, -1), (-1, -1), colors.HexColor('#f0fdf4')),
        ('GRID', (0, 0), (-1, -1), 0.5, colors.HexColor('#94a3b8')),
        ('ALIGN', (0, 0), (0, -1), 'CENTER'),
        ('ALIGN', (3, 0), (-1, -1), 'RIGHT'),
        ('TOPPADDING', (0, 0), (-1, -1), 3),
        ('BOTTOMPADDING', (0, 0), (-1, -1), 3),
    ]
    t.setStyle(TableStyle(t_style))
    elements.append(t)
    elements.append(Spacer(1, 8))

    # === TAX SUMMARY ===
    if inv.get("is_igst"):
        tax_rows = [
            ['Taxable Amount', f"Rs.{_fmt(totals.get('taxable', 0))}"],
            ['IGST', f"Rs.{_fmt(totals.get('igst', 0))}"],
            ['Grand Total', f"Rs.{_fmt(totals.get('total', 0))}"],
        ]
    else:
        tax_rows = [
            ['Taxable Amount', f"Rs.{_fmt(totals.get('taxable', 0))}"],
            ['CGST', f"Rs.{_fmt(totals.get('cgst', 0))}"],
            ['SGST', f"Rs.{_fmt(totals.get('sgst', 0))}"],
            ['Grand Total', f"Rs.{_fmt(totals.get('total', 0))}"],
        ]
    tax_t = RTable(tax_rows, colWidths=[120, 100])
    tax_t.setStyle(TableStyle([
        ('FONTNAME', (0, 0), (-1, -2), nfont), ('FONTNAME', (0, -1), (-1, -1), bfont),
        ('FONTSIZE', (0, 0), (-1, -1), 9), ('ALIGN', (1, 0), (1, -1), 'RIGHT'),
        ('TOPPADDING', (0, 0), (-1, -1), 2), ('BOTTOMPADDING', (0, 0), (-1, -1), 2),
        ('LINEBELOW', (0, -1), (-1, -1), 1, colors.black),
    ]))
    # Right-align the tax summary
    wrapper = RTable([['' , tax_t]], colWidths=[340, 220])
    wrapper.setStyle(TableStyle([('VALIGN', (0, 0), (-1, -1), 'TOP')]))
    elements.append(wrapper)
    elements.append(Spacer(1, 12))

    # === BANK DETAILS ===
    if company.get("bank_name") or company.get("bank_account"):
        elements.append(Paragraph("<b>Bank Details:</b>", ParagraphStyle('bk', fontName=bfont, fontSize=8)))
        bank_text = []
        if company.get("bank_name"):
            bank_text.append(f"Bank: {company['bank_name']}")
        if company.get("bank_account"):
            bank_text.append(f"A/C: {company['bank_account']}")
        if company.get("bank_ifsc"):
            bank_text.append(f"IFSC: {company['bank_ifsc']}")
        elements.append(Paragraph(" | ".join(bank_text), small_style))
        elements.append(Spacer(1, 8))

    # === NOTES ===
    if inv.get("notes"):
        elements.append(Paragraph(f"<b>Notes:</b> {inv['notes']}", small_style))
        elements.append(Spacer(1, 6))

    # === FOOTER ===
    elements.append(HRFlowable(width="100%", thickness=0.5, color=colors.HexColor('#cbd5e1')))
    elements.append(Spacer(1, 4))
    elements.append(Paragraph(f"Thank you - {company.get('company_name', '')}", ParagraphStyle('foot', fontName=nfont, fontSize=8, alignment=1, textColor=colors.grey)))

    doc.build(elements)
    buf.seek(0)
    fn = f"GST_Invoice_{inv.get('invoice_no', inv_id)}.pdf"
    return StreamingResponse(buf, media_type="application/pdf", headers={"Content-Disposition": f"attachment; filename={fn}"})

File: desktop-app/routes/gst_invoice.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const PDFDocument = require('pdfkit');
const path = require('path');
const fs = require('fs');

let registerFonts;
try {
  registerFonts = require('../shared/register_fonts');
} catch (_) {
  registerFonts = null;
}

module.exports = function(database) {
  const router = express.Router();
  const db = database;

  const safeSync = (fn) => (req, res, next) => { try { fn(req, res, next); } catch(e) { console.error('[GST_INVOICE]', e); res.status(500).json({ error: e.message }); }};
  const safeAsync = (fn) => async (req, res, next) => { try { await fn(req, res, next); } catch(e) { console.error('[GST_INVOICE]', e); res.status(500).json({ error: e.message }); }};

  // ============ GST COMPANY SETTINGS ============
  router.get('/api/gst-company-settings', safeSync((req, res) => {
    const settings = db.getData('/settings/gst_company', { company_name: '', gstin: '', address: '', state_code: '21', state_name: 'Odisha', phone: '', bank_name: '', bank_account: '', bank_ifsc: '' });
    res.json(settings);
  }));

  router.put('/api/gst-company-settings', safeSync((req, res) => {
    db.setData('/settings/gst_company', req.body);
    res.json({ success: true });
  }));

  // ============ GST INVOICE CRUD ============
  function calcTotals(items, is_igst) {
    const taxable = items.reduce((s, it) => s + (it.qty || 0) * (it.rate || 0), 0);
    const gst = items.reduce((s, it) => s + (it.qty || 0) * (it.rate || 0) * (it.gst_pct || 0) / 100, 0);
    return {
      taxable: Math.round(taxable * 100) / 100,
      gst: Math.round(gst * 100) / 100,
      cgst: is_igst ? 0 : Math.round(gst / 2 * 100) / 100,
      sgst: is_igst ? 0 : Math.round(gst / 2 * 100) / 100,
      igst: is_igst ? Math.round(gst * 100) / 100 : 0,
      total: Math.round((taxable + gst) * 100) / 100,
    };
  }

  router.get('/api/gst-invoices', safeSync((req, res) => {
    let invoices = db.getData('/gst_invoices', []);
    const { kms_year, season } = req.query;
    if (kms_year) invoices = invoices.filter(i => i.kms_year === kms_year);
    if (season) invoices = invoices.filter(i => i.season === season);
    invoices.sort((a, b) => (b.date || '').localeCompare(a.date || ''));
    res.json(invoices);
  }));

  router.post('/api/gst-invoices', safeSync((req, res) => {
    const inv = { ...req.body, id: uuidv4(), created_at: new Date().toISOString() };
    inv.totals = calcTotals(inv.items || [], inv.is_igst);
    const invoices = db.getData('/gst_invoices', []);
    invoices.push(inv);
    db.setData('/gst_invoices', invoices);
    res.json(inv);
  }));

  router.put('/api/gst-invoices/:id', safeSync((req, res) => {
    const invoices = db.getData('/gst_invoices', []);
    const idx = invoices.findIndex(i => i.id === req.params.id);
    if (idx === -1) return res.status(404).json({ error: 'Not found' });
    const upd = { ...req.body, id: req.params.id, created_at: invoices[idx].created_at, updated_at: new Date().toISOString() };
    upd.totals = calcTotals(upd.items || [], upd.is_igst);
    invoices[idx] = upd;
    db.setData('/gst_invoices', invoices);
    res.json({ success: true, totals: upd.totals });
  }));

  router.delete('/api/gst-invoices/:id', safeSync((req, res) => {
    let invoices = db.getData('/gst_invoices', []);
    const before = invoices.length;
    invoices = invoices.filter(i => i.id !== req.params.id);
    if (invoices.length === before) return res.status(404).json({ error: 'Not found' });
    db.setData('/gst_invoices', invoices);
    res.json({ success: true });
  }));

  // ============ GST INVOICE PDF ============
  router.get('/api/gst-invoices/:id/pdf', safeSync((req, res) => {
    const invoices = db.getData('/gst_invoices', []);
    const inv = invoices.find(i => i.id === req.params.id);
    if (!inv) return res.status(404).json({ error: 'Not found' });

    const company = db.getData('/settings/gst_company', {});
    const doc = new PDFDocument({ size: 'A4', margin: 40 });

    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename=GST_Invoice_${inv.invoice_no || inv.id}.pdf`);
    doc.pipe(res);

    if (registerFonts) registerFonts(doc);
    const bfont = doc._registeredFonts && doc._registeredFonts['FreeSansBold'] ? 'FreeSansBold' : 'Helvetica-Bold';
    const nfont = doc._registeredFonts && doc._registeredFonts['FreeSans'] ? 'FreeSans' : 'Helvetica';

    const W = doc.page.width - 80; // total usable width
    const L = 40; // left margin

    // === HEADER ===
    doc.font(bfont).fontSize(16).text(company.company_name || 'COMPANY NAME', L, 40, { align: 'center', width: W });
    let y = doc.y;
    if (company.address) { doc.font(nfont).fontSize(9).fillColor('#666').text(company.address, L, y, { align: 'center', width: W }); y = doc.y; }
    if (company.gstin) { doc.font(nfont).fontSize(9).fillColor('#666').text(`GSTIN: ${company.gstin}`, L, y, { align: 'center', width: W }); y = doc.y; }
    if (company.phone) { doc.font(nfont).fontSize(9).fillColor('#666').text(`Phone: ${company.phone}`, L, y, { align: 'center', width: W }); y = doc.y; }
    y += 6;
    doc.moveTo(L, y).lineTo(L + W, y).strokeColor('#334155').lineWidth(1).stroke();
    y += 8;

    // TAX INVOICE
    doc.font(bfont).fontSize(14).fillColor('#1e40af').text('TAX INVOICE', L, y, { align: 'center', width: W });
    y = doc.y + 8;

    // Invoice info
    doc.font(nfont).fontSize(9).fillColor('#000');
    doc.text(`Invoice No: ${inv.invoice_no || ''}`, L, y);
    doc.text(`Date: ${inv.date || ''}`, L + W - 150, y, { width: 150, align: 'right' });
    y = doc.y + 4;
    doc.font(bfont).fontSize(10).text(`Bill To: ${inv.buyer_name || ''}`, L, y);
    y = doc.y + 2;
    doc.font(nfont).fontSize(9);
    if (inv.buyer_gstin) { doc.text(`GSTIN: ${inv.buyer_gstin}`, L, y); y = doc.y + 1; }
    if (inv.buyer_address) { doc.text(`Address: ${inv.buyer_address}`, L, y); y = doc.y + 1; }
    y += 8;

    // === ITEMS TABLE ===
    const cols = [25, 120, 65, 45, 40, 55, 65, 35, 55, 60]; // ~565 total
    const headers = ['#', 'Item', 'HSN', 'Qty', 'Unit', 'Rate', 'Taxable', 'GST%', 'GST Amt', 'Total'];
    const rowH = 18;

    // Header row
    let x = L;
    doc.rect(L, y, W, rowH).fill('#e2e8f0');
    doc.font(bfont).fontSize(7).fillColor('#000');
    headers.forEach((h, i) => {
      const align = i >= 3 ? 'right' : 'left';
      doc.text(h, x + 2, y + 4, { width: cols[i] - 4, align });
      x += cols[i];
    });
    y += rowH;

    // Data rows
    const items = inv.items || [];
    const fmt = (v) => v ? v.toLocaleString('en-IN', { minimumFractionDigits: 2, maximumFractionDigits: 2 }) : '0.00';
    doc.font(nfont).fontSize(7).fillColor('#000');
    items.forEach((it, idx) => {
      const taxable = (it.qty || 0) * (it.rate || 0);
      const gst_a = taxable * (it.gst_pct || 0) / 100;
      const total = taxable + gst_a;
      const row = [String(idx + 1), it.name || '', it.hsn || '', String(it.qty || 0), it.unit || '', `Rs.${fmt(it.rate)}`, `Rs.${fmt(taxable)}`, `${it.gst_pct || 0}%`, `Rs.${fmt(gst_a)}`, `Rs.${fmt(total)}`];
      if (idx % 2 === 1) doc.rect(L, y, W, rowH).fill('#f8fafc').fillColor('#000');
      x = L;
      row.forEach((val, i) => {
        const align = i >= 3 ? 'right' : 'left';
        doc.text(val, x + 2, y + 4, { width: cols[i] - 4, align });
        x += cols[i];
      });
      y += rowH;
    });

    // Totals row
    const totals = inv.totals || {};
    doc.rect(L, y, W, rowH).fill('#f0fdf4');
    doc.font(bfont).fontSize(7).fillColor('#000');
    x = L;
    ['', '', '', '', '', '', `Rs.${fmt(totals.taxable)}`, '', `Rs.${fmt(totals.gst)}`, `Rs.${fmt(totals.total)}`].forEach((val, i) => {
      doc.text(val, x + 2, y + 4, { width: cols[i] - 4, align: i >= 3 ? 'right' : 'left' });
      x += cols[i];
    });
    y += rowH;

    // Grid lines
    doc.strokeColor('#94a3b8').lineWidth(0.5);
    const tableTop = y - rowH * (items.length + 2);
    for (let r = 0; r <= items.length + 2; r++) {
      const ry = tableTop + r * rowH;
      doc.moveTo(L, ry).lineTo(L + W, ry).stroke();
    }
    x = L;
    for (let c = 0; c <= cols.length; c++) {
      doc.moveTo(x, tableTop).lineTo(x, y).stroke();
      x += (cols[c] || 0);
    }
    y += 10;

    // === TAX SUMMARY ===
    doc.font(nfont).fontSize(9).fillColor('#000');
    const taxX = L + W - 220;
    doc.text('Taxable Amount:', taxX, y); doc.text(`Rs.${fmt(totals.taxable)}`, taxX + 120, y, { width: 100, align: 'right' }); y += 14;
    if (inv.is_igst) {
      doc.text('IGST:', taxX, y); doc.text(`Rs.${fmt(totals.igst)}`, taxX + 120, y, { width: 100, align: 'right' }); y += 14;
    } else {
      doc.text('CGST:', taxX, y); doc.text(`Rs.${fmt(totals.cgst)}`, taxX + 120, y, { width: 100, align: 'right' }); y += 14;
      doc.text('SGST:', taxX, y); doc.text(`Rs.${fmt(totals.sgst)}`, taxX + 120, y, { width: 100, align: 'right' }); y += 14;
    }
    doc.moveTo(taxX, y).lineTo(taxX + 220, y).strokeColor('#000').lineWidth(1).stroke(); y += 4;
    doc.font(bfont).fontSize(11).text('Grand Total:', taxX, y); doc.text(`Rs.${fmt(totals.total)}`, taxX + 120, y, { width: 100, align: 'right' }); y += 20;

    // Bank Details
    if (company.bank_name || company.bank_account) {
      doc.font(bfont).fontSize(8).text('Bank Details:', L, y); y += 12;
      doc.font(nfont).fontSize(8);
      const parts = [];
      if (company.bank_name) parts.push(`Bank: ${company.bank_name}`);
      if (company.bank_account) parts.push(`A/C: ${company.bank_account}`);
      if (company.bank_ifsc) parts.push(`IFSC: ${company.bank_ifsc}`);
      doc.text(parts.join(' | '), L, y); y += 16;
    }

    // Notes
    if (inv.notes) {
      doc.font(bfont).fontSize(8).text('Notes:', L, y); y += 12;
      doc.font(nfont).fontSize(8).text(inv.notes, L, y); y += 16;
    }

    // Footer
    doc.moveTo(L, y).lineTo(L + W, y).strokeColor('#cbd5e1').lineWidth(0.5).stroke(); y += 6;
    doc.font(nfont).fontSize(8).fillColor('#999').text(`Thank you - ${company.company_name || ''}`, L, y, { align: 'center', width: W });

    doc.end();
  }));

  return router;
};
